ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/PartitionByParamsFactory.java

```diff
@@ -0,0 +1,190 @@
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.util.EngineProcessingLogMessageFactory;
+import io.confluent.ksql.execution.util.ExpressionTypeManager;
+import io.confluent.ksql.execution.util.StructKeyUtil;
+import io.confluent.ksql.execution.util.StructKeyUtil.KeyBuilder;
+import io.confluent.ksql.function.FunctionRegistry;
+import io.confluent.ksql.logging.processing.ProcessingLogger;
+import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.name.ColumnNames;
+import io.confluent.ksql.schema.ksql.Column;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.LogicalSchema.Builder;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.KsqlConfig;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.streams.KeyValue;
+
+public final class PartitionByParamsFactory {
+
+  private PartitionByParamsFactory() {
+  }
+
+  public static PartitionByParams build(
+      final LogicalSchema sourceSchema,
+      final Expression partitionBy,
+      final KsqlConfig ksqlConfig,
+      final FunctionRegistry functionRegistry,
+      final ProcessingLogger logger
+  ) {
+    final Optional<Column> partitionByCol = getPartitionByCol(sourceSchema, partitionBy);
+
+    final Function<GenericRow, Object> evaluator = buildExpressionEvaluator(
+        sourceSchema,
+        partitionBy,
+        ksqlConfig,
+        functionRegistry,
+        logger
+    );
+
+    final LogicalSchema resultSchema =
+        buildSchema(sourceSchema, partitionBy, functionRegistry, partitionByCol);
+
+    return buildMapper(resultSchema, partitionByCol, evaluator);
+  }
+
+  public static LogicalSchema buildSchema(
+      final LogicalSchema sourceSchema,
+      final Expression partitionBy,
+      final FunctionRegistry functionRegistry
+  ) {
+    final Optional<Column> partitionByCol = getPartitionByCol(sourceSchema, partitionBy);
+    return buildSchema(sourceSchema, partitionBy, functionRegistry, partitionByCol);
+  }
+
+  private static LogicalSchema buildSchema(
+      final LogicalSchema sourceSchema,
+      final Expression partitionBy,
+      final FunctionRegistry functionRegistry,
+      final Optional<Column> partitionByCol
+  ) {
+    final ExpressionTypeManager expressionTypeManager =
+        new ExpressionTypeManager(sourceSchema, functionRegistry);
+
+    final SqlType keyType = expressionTypeManager
+        .getExpressionSqlType(partitionBy);
+
+    final ColumnName newKeyName = partitionByCol
+        .map(Column::name)
+        .orElseGet(() -> ColumnNames.nextGeneratedColumnAlias(sourceSchema));
+
+    final Builder builder = LogicalSchema.builder()
+        .withRowTime()
+        .keyColumn(newKeyName, keyType)
+        .valueColumns(sourceSchema.value());
```

Comment:
clearly this happens somewhere, but I'm not sure where, don't we need to remove the `paritionByCol` if it existed in the value?

Reply:
No, we don't need to remove it, cos we need it there.

Consider:

A => B, C

And applying PARTITION BY B.

The final schema should be:

B => C, A

However, during processing we copy the key columns into the value, hence the source schema is:

A => B, C, A

When we repartition this becomes:

B => B, C, A

Notice, no need to change the value!

However, if PARTITION BY UDF(B), then:

A => B, C

Becomes:

KSQL_COL_0 => B, C, A

Which during processing is actually:

A => B, C, A

Becoming

KSQL_COL_0 => B, C, A, KSQL_COL_0.

This would all likely be a lot clearer if we had a `LogicalSchema` and a `ProcessingSchema` type, but ... that would be a lot of work!

Reply:
I've added Java docs to cover this.